### PR TITLE
LocalStore::addTempRoot(): Handle ENOENT

### LINKED ISSUE #7370: Crash during concurrent garbage collection
### Description

`nix` builds intermittently fail with the following error message:

```
error: cannot connect to socket at '/nix/var/nix/gc-socket/socket': No such file or directory
```

This only appears to occur on systems with scheduled garbage collection and/or concurrent calls to collect garbage.

**UPDATE (29 March 2023)**: In addition to the above failure, we have also been seeing the following errors occur in close proximity to scheduled garbage collection; it's not entirely clear _why_ Nix might GC a store path that's a dependency of an active realization, but that does appear to be what's happening:

```
FileNotFoundError: [Errno 2] No such file or directory: '/nix/store/<NIX HASH>-<TARGET>'
```

### Reproduction

Unable to consistently reproduce.

### Expected Behavior

Nix should be able to do one of the following:
- perform concurrent garbage collection without crashing
- make concurrent garbage collection optional

### Nix Version

```
❯ nix --version
nix (Nix) 2.10.3
```

#### Related Issues & Code Paths

AFAICT this seems to have been introduced after #5149; at the very least, this problem did not seem to appear on systems running Nix 2.4 & has since appeared with some frequency on systems running Nix 2.10.

The error appears to be coming from this code path: https://github.com/NixOS/nix/blob/0f64bf445a229c4f5ce19dc19a552f58701ca1ca/src/libstore/gc.cc#L118-L121

#### Potential Causes

This is very much outside of my area of expertise, but my guess is that the following occurs:

- process A starts GC, takes the lock
- process B starts GC, sees that there is a lock, and attempts to communicate with the GC process via its GC socket
- process A releases the lock **before** process B can attempt to probe for the socket
- process B fails

I don't know that this matters at all, but looking at the following code block I would have expected Nix to call the GC socket with `send`/`recv` rather than `writeFull`/`readFull`:
https://github.com/NixOS/nix/blob/0f64bf445a229c4f5ce19dc19a552f58701ca1ca/src/libstore/gc.cc#L140-L161

### PR DESCRIPTION
**Pull request by edolstra**
# Motivation
If the garbage collector has acquired the global GC lock, but hasn't created the GC socket yet, then a client attempting to connect would get ENOENT. Note that this only happens when the GC runs for the first time on a machine. Subsequently clients will get ECONNREFUSED which was already handled.

Fixes #7370.

# Context

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: src/libstore/gc.cc [hunk 1]
```diff
@@ -138,9 +138,9 @@ void LocalStore::addTempRoot(const StorePath & path)
             try {
                 nix::connect(fdRootsSocket->get(), socketPath);
             } catch (SysError & e) {
-                /* The garbage collector may have exited, so we need to
-                   restart. */
-                if (e.errNo == ECONNREFUSED) {
+                /* The garbage collector may have exited or not
+                   created the socket yet, so we need to restart. */
+                if (e.errNo == ECONNREFUSED || e.errNo == ENOENT) {
                     debug("GC socket connection refused");
```
Comment (Ericson2314) [inline src/libstore/gc.cc:144]: Good to know which one it was
```suggestion
                    debug("GC socket connection refused: %s", e.msg())
```

**Comment by Ericson2314** [comment]
@edolstra should we do `connect` too?

**Comment by Ericson2314** [comment]
Did `bind` in #9787

**Comment by GrahamDennis** [comment]
Backport to Nix-2.18 here: https://github.com/NixOS/nix/pull/13896. There was an attempt by @roberth to backport this to 2.18 and 2.15 above (https://github.com/NixOS/nix/pull/8544#issuecomment-1894560751) but that failed.

**Pull request merged**